components/director/internal/domain/scenarioassignment/engine.go

```diff
@@ -15,35 +13,70 @@ import (
 //go:generate mockery -name=LabelRepository -output=automock -outpkg=automock -case=underscore
 type LabelRepository interface {
 	GetRuntimeScenariosWhereLabelsMatchSelector(ctx context.Context, tenantID, selectorKey, selectorValue string) ([]model.Label, error)
-	Upsert(ctx context.Context, label *model.Label) error
+	GetRuntimesIDsByKeyAndValue(ctx context.Context, tenantID, selectorKey, selectorValue string) ([]string, error)
+	GetScenarioLabelsForRuntimes(ctx context.Context, tenantID string, runtimesIDs []string) ([]model.Label, error)
+	Delete(ctx context.Context, tenant string, objectType model.LabelableObject, objectID string, key string) error
+}
+
+//go:generate mockery -name=LabelUpsertService -output=automock -outpkg=automock -case=underscore
+type LabelUpsertService interface {
+	UpsertLabel(ctx context.Context, tenant string, labelInput *model.LabelInput) error
 }
 
 type engine struct {
 	labelRepo              LabelRepository
 	scenarioAssignmentRepo Repository
+	labelService           LabelUpsertService
 }
 
-func NewEngine(labelRepo LabelRepository, scenarioAssignmentRepo Repository) *engine {
+func NewEngine(labelService LabelUpsertService, labelRepo LabelRepository, scenarioAssignmentRepo Repository) *engine {
 	return &engine{
 		labelRepo:              labelRepo,
 		scenarioAssignmentRepo: scenarioAssignmentRepo,
+		labelService:           labelService,
 	}
 }
 
 func (e *engine) EnsureScenarioAssigned(ctx context.Context, in model.AutomaticScenarioAssignment) error {
-	labels, err := e.labelRepo.GetRuntimeScenariosWhereLabelsMatchSelector(ctx, in.Tenant, in.Selector.Key, in.Selector.Value)
+	runtimesIDs, err := e.labelRepo.GetRuntimesIDsByKeyAndValue(ctx, in.Tenant, in.Selector.Key, in.Selector.Value)
 	if err != nil {
-		return errors.Wrap(err, "while getting runtimes scenarios which match given selector")
+		return errors.Wrapf(err, "while fetching runtimes id which match given selector:%+v", in)
+	}
+
+	if len(runtimesIDs) == 0 {
+		return nil
 	}
-	return e.upsertMergedScenarios(ctx, labels, in.ScenarioName, e.uniqueScenarios)
+
+	labels, err := e.labelRepo.GetScenarioLabelsForRuntimes(ctx, in.Tenant, runtimesIDs)
+	if err != nil {
+		return errors.Wrap(err, "while fetching scenarios labels for matched runtimes")
+	}
+	err = e.upsertMergedScenarios(ctx, labels, in.Tenant, in.ScenarioName, e.uniqueScenarios)
+	if err != nil {
+		return errors.Wrap(err, "while upserting merged scenarios to runtimes")
+	}
+
+	rtmWithoutScenarios := make(map[string]interface{})
```

Comment:
Hmm I wonder why we need to iterate again through runtimes without scenarios?
Can we just iterate once for all runtimes and upsert scenario labels for them?